src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UnliftedNewtypes #-}
+
+-- |
+-- This module provides an unlifted mutable array with a pure
+-- interface. Though the array itself is unlifted, it's elements are
+-- lifted types. This is made possible by using linear types to make
+-- sure array references are single threaded through reads and writes.
+--
+-- Accessing out-of-bounds indices causes undefined behaviour.
+--
+-- This module is meant to be imported qualified.
+module Data.Array.Mutable.Unlifted.Linear
+  ( Array#
+  , unArray#
+  , alloc
+  , allocBeside
+  , lseq
+  , size
+  , read
+  , write
+  , copyInto
+  ) where
+
+import Data.Unrestricted.Linear hiding (lseq)
+import Prelude (Int)
+import qualified Prelude as Prelude
+import qualified Unsafe.Linear as Unsafe
+import qualified GHC.Exts as GHC
+
+-- | A mutable array holding @a@s
+newtype Array# a = Array# (GHC.MutableArray# GHC.RealWorld a)
+
+-- | Extract the underlying 'GHC.MutableArray#', consuming the 'Array#'
+-- in process.
+unArray# :: (GHC.MutableArray# GHC.RealWorld a -> b) -> Array# a #-> Ur b
+unArray# f = Unsafe.toLinear (\(Array# a) -> Ur (f a))
+
+-- | Consume an 'Array#'.
+--
+-- Note that we can not implement a 'Consumable' instance because 'Array#'
+-- is unlifted.
+lseq :: Array# a #-> b #-> b
+lseq = Unsafe.toLinear2 (\_ b -> b)
+
+-- | Allocate a mutable array of given size using a default value.
+--
+-- The size should be non-negative.
+alloc :: Int -> a -> (Array# a #-> Ur b) #-> Ur b
+alloc (GHC.I# s) a f =
+  let new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in f new
+{-# NOINLINE alloc #-}
+
+-- For the reasoning behind these NOINLINE pragmas, see the discussion at:
+-- https://github.com/tweag/linear-base/pull/187#pullrequestreview-489183531
+
+-- | Allocate a mutable array of given size using a default value,
+-- using another 'Array#' as a uniqueness proof.
+--
+-- The size should be non-negative.
+allocBeside :: Int -> a -> Array# b #-> (# Array# b, Array# a #)
+allocBeside (GHC.I# s) a orig =
+  let new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in (# orig, new #)
+{-# NOINLINE allocBeside #-}
+
+size :: Array# a #-> (# Array# a, Ur Int #)
+size = Unsafe.toLinear go
+  where
+    go :: Array# a -> (# Array# a, Ur Int #)
+    go (Array# arr) =
+      let !s = GHC.sizeofMutableArray# arr
+      in  (# Array# arr, Ur (GHC.I# s) #)
+
+read ::  Int -> Array# a #-> (# Array# a, Ur a #)
+read (GHC.I# i) = Unsafe.toLinear go
```

Comment:
Okay, the rough argument for why this works needs to be documented in /docs/DESIGN.md. 

Here's my understanding of it:

For mutable data to work properly in Haskell, the constructing, mutating and accessing operations need to atomic. That does not mean that they are used strictly. We can have a lazy `(read arr ix)`. We just require that any evaluation of `(read arr ix)` fully does the read (even if the value read is itself lazy). For constructors and mutators that return a single argument, this is already done by necessity. Evaluating `((f :: Array Int #-> Ur Bool) newArray)` forces the array to be created before any writes or reads are done to it. Similarly, any computation that uses a `(write arr ix a) & \arr' -> (...)` must do the write before reading from the written array.

The issue arises when we have API functions that return multiple things like `read`. Here, when we evaluate `(read arr ix)` and get out the read value and the array, we must never have a thunk of the read value that has yet to perform an mutable read of the underlying array. The same goes for a copying function like `slice`. We should never be able to use the first part of `(slice arr ix n)` (that is, the original array) without having the slice be done to the mutable array underneath.

Here, we can reason as follows. We want to show that pattern matching on `read arr ix` forces the underlying read to occur even if we just discard it. First, note that when `(read arr ix)` is evaluated, it is done without inlining. Thus, we only need to argue that local optimizations within this functions won't mess things up. 

Now, when matching on unlifted tuples, we can be sure what's held inside is also unlifted. That is, the values cannot be bottom. If `Ur ret` is not bottom, then (I think) `ret` is not bottom. For this to be the case, GHC must evaluate the `GHC.runRW#  (GHC.readArray# arr i)`. Even if GHC re-writes this to `(# Array# arr, Ur (GHC.runRW# ...))` it must not leave the thunk lest it be bottom. Hence, the underlying read occurs on the pattern match and `read` is atomic.

@facundominguez @aspiwack Is this argument clear and correct?



A mutator or constructor like `write` or `newArray` will never cause issues because to do anything with their result, they must be evaluated first.

Reply:
> mutating and accessing operations need to be atomic.

I'm guessing atomic here means that the compiler can't inline them and mix their subexpressions? I need to ask because I've only seen atomic used to describe transactions and concurrent operations.

> Now, when matching on unlifted tuples, we can be sure what's held inside is also unlifted. 

This is true when the type of the component is unlifted.
```Haskell
case (# Array# arr, undefined #) of
  (# arr, d #) -> ()  -- this should force the first component of the pair only.
```

Reply:
I think that this explanation isn't quite accurate.

The requirement that you have is rather that, outside the definition of primitives, there is no aliasing of the array. It means, in particular, that if you return the array and something else, then the something else must never point in any way to the array. Otherwise, that something could observe a subsequent mutation.

It's pretty much the same story as in Rust, here.

Reply:
Oh ... right! And that's why we need the `(GHC.runRW# ...)` thing to be evaluated when the `(read arr ix)` is evaluated. More pointedly then, is my argument for why the read is evaluated sound? From what @facundominguez pointed out, since  `(GHC.runRW# ...)` has a lifted type, it might not hold. In that case, are we sure that `case (GHC.runRW# ...) of (_, a) -> (Array arr, a)` can never be rewritten to `(Array arr, case (GHC.runRW# ...) of (_, a) -> a)`? If so, then we're still in the green zone.